src/client/cl_main.cpp

```diff
@@ -4043,3 +4046,128 @@ void CL_GetVMGLConfig(vmglconfig_t *vmglconfig) {
 	vmglconfig->stereoEnabled = cls.glconfig.stereoEnabled;
 	vmglconfig->smpActive = cls.glconfig.smpActive;
 }
+
+void CL_ShaderStateChanged( void ) {
+	// Originally copied from CG_ShaderStateChanged, but rewritten to avoid issues of the original implementation and to
+	// support settings
+	char originalShader[MAX_QPATH];
+	char newShader[MAX_QPATH];
+	char settings[32];
+	char *timeOffset;
+	char *lightmapMode;
+	char *styleMode;
+
+	shaderRemapLightmapType_t lightmapModeValue;
+	shaderRemapStyleType_t styleModeValue;
+
+	const char *curPos = cl.gameState.stringData + cl.gameState.stringOffsets[ CS_MV_REMAPS ];
+	const char *endPos = curPos + strlen( curPos );
+
+	// Check configstring for prefix to avoid misinterpreting configstrings from mods
+	if ( !curPos || !*curPos || strncmp(curPos, "mvremap:", 8) ) return;
+
+	// Handling of these is really ugly. I originally wanted to handle them like the base game/cgame modules do, but
```

Comment:
I think in a string parsing function some examples and syntax explanation would be helpful.

Reasoning is ok, but once we commit it there is no changing the syntax anyway.

Reply:
Do you think an example like this would be helpful: `6console4clear929300;p;p;`:
 - `6` -> ascii 54 -> 54 - 47 = 7 -> the next 7 characters are the source shader name
 - `console` -> source shader (7 characters)
 - `4` -> ascii 52 -> 52 - 47 = 5 -> the next 5 characters are the destination shader name
 - `clear` -> destination shader (5 characters)
 - `9` -> ascii 57 -> 57 - 47 = 10 -> the next 10 characters are the options
 - `29300;p;p;`
   - `29300` -> timeOffset
   - `p` -> lightmapMode
   - `p` -> styleMode
  
Here is a simple adaption of `BuildShaderStateConfig` that uses the existing base remap struct and array to build mv remaps with lightmapMode and styleMode set to preserve. Calculating timeOffset `* 1000`, because as of fe717366f617d4645db0540ee714e971345b81d3 the remaps expect to get the offset as millisecond integer. Proper mods should just skip the `* 0.001f` step when adding remaps and instead apply the conversion when building the legacy shader state.
```c
const char *BuildShaderStateConfigMV(void)
{
    static char buff[BIG_INFO_VALUE];
    char settings[32];
    int i;

    buff[0] = 0;
    for ( i = 0; i < remapCount; i++ )
    {
        Com_sprintf( settings, sizeof(settings), "%i;p;p;", remappedShaders[i].timeOffset * 1000 );
        Q_strcat( buff, sizeof(buff), va("%c%s%c%s%c%s", strlen(remappedShaders[i].oldShader)+47, remappedShaders[i].oldShader, strlen(remappedShaders[i].newShader)+47, remappedShaders[i].newShader, strlen(settings)+47, settings) );
    }
    return buff;
}
```

Example to set it:
```c
// Prepend the `mvremap:` prefix so jk2mv clients know this is supposed to be a configstring holding mvremaps and not some accidental collision with other content. This prefix could be set within BuildShaderStateConfigMV to avoid this extra call to `va`.
trap_SetConfigstring( 1399, va("mvremap:%s", BuildShaderStateConfigMV()) );
```

Reply:
Example looks great. I would just add mvremap: at the begining and say you can append more afterwards for full picture of configstring being parsed. module examples we can add in the wiki - there isn't really a system do document module syscalls in engine code sadly.